remotes/docker/resolver.go

```diff
@@ -148,7 +148,11 @@ func NewResolver(options ResolverOptions) remotes.Resolver {
 
 	if options.Headers == nil {
 		options.Headers = make(http.Header)
+	} else {
+		// make a copy of the headers to avoid race due to concurrent map write
+		options.Headers = options.Headers.Clone()
```

Comment:
Slightly wondering if we're tapering over a larger problem here; if `options` are passed by reference, and  `options.Headers` is mutated concurrently, wouldn't the same apply to any of the other fields in `options` ?

Reply:
Yeah I think there _might_ be a larger problem, depending on how various `Options` are used (i.e. any in-place modify) and types of fields modified.

It's not `options` as a whole being passed by reference, it's the underlying `map`-like fields (or `slice`, `pointer`) in it. For other types, it should be fine to modify even in place within a func since they have their own copy.

For `map/slice/pointer`-like types, if we make in-place modify in a func and avoid race, we either copy while passing it to the func, or copy within the func (like in this fix).

Reply:
not sure if there is a good way of checking/finding this kind of unexpected data race (i.e.  due to `map/slice/pointer` sharing the same underlying data) other than `-race` 🤔️

Reply:
> It's not options as a whole being passed by reference, it's the underlying map-like fields (or slice, pointer) in it.

Good point, I realised that part after I commented 👍 


But indeed, there may be other issues hiding here 🤔 (if not in containerd's source code, then still possibly in other projects using this package)

Very orthogonal, but I was actually looking at this part of the code recently, as it's somewhat awkward to use (there's a lot of code inside this function, part of which is to provide the backward-compatibility for the deprecated options), perhaps some code should be extracted from this at some point.

Reply:
I've always found this design kind of odd (NewResolver), as for CRI if you look at image_pull, we need to make a new Resolver on every request because it needs the auth info upfront. I'd think it'd make more sense to create one resolver object we can supply up front and on use of the object (Resolve or other method) supply auth/hosts info, but I never dug deep enough to say "okay we can", this was just surface level thoughts. At least in that case the CRI issue would be fixed just by API changes

Reply:
Wouldn't the use of `options.Headers.Clone()` still subject to concurrent read and write in the following scenario?

* goroutine_1: doing `options.Headers.Clone()` while goroutine_2 is doing `options.Headers.Set("User-Agent", "containerd/"+version.Version)`, at this moment the code here read and write to the `options.Headers` concurrently. 

![image](https://github.com/containerd/containerd/assets/28563006/a1ffae4e-cb96-444e-87a2-6b2fce28c728)

Should we add a RWLock to the `ResolverOptions` and protect the map access with the lock?

Reply:
I don't know if it'd work if it was a ResolverOption, you'd have to provide the same lock to N different ResolverOptions so CRI would have to keep around some mutex just to pass into all NewResolver calls; I don't think a client should have to provide a/the mutex. 

The simplest solution is probably just a `var headersMu sync.Mutex` global right above NewResolver we can use, so the same lock is used on any uses of NewResolver. Or we can clone the headers in the CRI plugin and document NewResolver isn't thread safe when using the same headers object 😐.

Reply:
@shuaichang I don't think it will be a race. `ResolverOptions` is a struct so every `NewResolver` call will has its own copy of `ResolverOptions`. The previous race is due to multiple copy pointing to the same underlying map/header. (i.e. `options.Headers` as a field is not shared, but the map it points to is shared, similar to slice/pointer). With the change, each copy of `options.Headers` points to either a new empty Header (line 150) or a cloned Header (line 153). Basically `NewResolver` now only reads the original `header` passed to it (previously it read AND write).

As an example, the following code works:

```golang
options := ResolverOptions{Headers: make(http.Header)}
go NewResolver(options)
go NewResolver(options)
```

The only race is the caller itself is somehow modifying the header (now it becomes a read-write race). But in that case personally I think it's the caller's responsibility to ensure it doesn't have such read-write race.

Reply:
can we use the following example to explain if the concurrent read/write could happen to the underlaying map, t1, t2 means timing, gr1, gr2 means goroutine

* t1: 
  * gr1: calls docker.NewResolver(header=Nil)
  * gr2: calls docker.NewResolver(header=Nil)
* t2:
  * gr1: options.Headers = make(http.Header), Headers points to map1
  * gr2: can be at arbitrary code line before checking if options.Headers == nil {
* t3:
  * gr1: options.Headers.Set("User-Agent", "containerd/"+version.Version), this writes to map1
  * gr2: options.Headers = options.Headers.Clone(), this reads from map1

So my understanding is t3 there will be a read/write race, unless I missed something.

Reply:
ah, nvm, I think my example ^, t3 gr2 will not reach the Clone line as we passed the options as a struct copy.